stl/src/tzdb.cpp

```diff
@@ -1,15 +1,345 @@
+    }
+
+    _NODISCARD const char* _Allocate_wide_to_narrow(
+        const char16_t* _Input, int _Input_len, __std_tzdb_error& _Err) noexcept {
+        const auto _Code_page      = __std_fs_code_page();
+        const auto _Input_as_wchar = reinterpret_cast<const wchar_t*>(_Input);
+        const auto _Count_result = __std_fs_convert_wide_to_narrow(_Code_page, _Input_as_wchar, _Input_len, nullptr, 0);
+        if (_Count_result._Err != __std_win_error::_Success) {
+            _Err = __std_tzdb_error::_Win_error;
+            return nullptr;
+        }
+
+        _STD unique_ptr<char[]> _Data{new (_STD nothrow) char[_Count_result._Len + 1]};
+        if (_Data == nullptr) {
+            return nullptr;
+        }
+
+        _Data[_Count_result._Len] = '\0';
+
+        const auto _Result =
+            __std_fs_convert_wide_to_narrow(_Code_page, _Input_as_wchar, _Input_len, _Data.get(), _Count_result._Len);
+        if (_Result._Err != __std_win_error::_Success) {
+            _Err = __std_tzdb_error::_Win_error;
+            return nullptr;
+        }
+
+        return _Data.release();
+    }
+
+    _NODISCARD _STD unique_ptr<char16_t[]> _Get_canonical_id(
+        const char16_t* _Id, int32_t _Len, int32_t& _Result_len, __std_tzdb_error& _Err) {
+        static constexpr int32_t _Link_buf_len = 32;
+        _STD unique_ptr<char16_t[]> _Link_buf{new (_STD nothrow) char16_t[_Link_buf_len]};
+        if (_Link_buf == nullptr) {
+            return nullptr;
+        }
+
+        UErrorCode _UErr{U_ZERO_ERROR};
+        UBool _Is_system{};
+        _Result_len = __icu_ucal_getCanonicalTimeZoneID(_Id, _Len, _Link_buf.get(), _Link_buf_len, &_Is_system, &_UErr);
+        if (_UErr == U_BUFFER_OVERFLOW_ERROR && _Result_len > 0) {
+            _Link_buf.reset(new (_STD nothrow) char16_t[_Result_len + 1]);
+            if (_Link_buf == nullptr) {
+                return nullptr;
+            }
+
+            _UErr = U_ZERO_ERROR; // reset error.
+            _Result_len =
+                __icu_ucal_getCanonicalTimeZoneID(_Id, _Len, _Link_buf.get(), _Result_len, &_Is_system, &_UErr);
+            if (U_FAILURE(_UErr)) {
+                _Err = __std_tzdb_error::_Icu_error;
+                return nullptr;
+            }
+        } else if (U_FAILURE(_UErr) || _Result_len <= 0) {
+            _Err = __std_tzdb_error::_Icu_error;
+            return nullptr;
+        }
+
+        return _Link_buf;
+    }
+
+    template <class _Ty, class _Dtor>
+    _NODISCARD constexpr _Ty* _Report_error(_STD unique_ptr<_Ty, _Dtor>& _Info, __std_tzdb_error _Err) {
+        _Info->_Err = _Err;
+        return _Info.release();
+    }
+
+    template <class _Ty, class _Dtor>
+    _NODISCARD constexpr _Ty* _Propagate_error(_STD unique_ptr<_Ty, _Dtor>& _Info) {
+        // a bad_alloc() returns nullptr and does not set __std_tzdb_error
+        return _Info->_Err != __std_tzdb_error::_Success ? _Info.release() : nullptr;
+    }

```

Comment:
As `unique_ptr` isn't `constexpr`, I believe that `constexpr` on these functions should be dropped.